util/functools.py

```diff
@@ -0,0 +1,389 @@
+"""A module created to allow similar functionality to the builtin functools, but addressing a few limitations."""
```

Comment:
I actually created this for another pet project, but I find it pretty useful as an alternative to the builtin functools. The reason this exists is to allow single dispatch with the following:

- `from __future__ import annotations` issues where a method/function defined within a class that uses a type hint for that class was not supported by the built in functools (at least in our version of python). 
- The use of the first argument provided as a key word argument during call. 

I think you are better off using the built in annotations if these two issues don't impact you, but they are available incase you need them.